File: qd_hrms_app/qd_hrms/public/js/self_service.js
(() => {
	const CONTACT_FIELDS = new Set([
		"image",
		"cell_number",
		"personal_email",
		"prefered_contact_email",
		"prefered_email",
		"current_address",
		"current_accommodation_type",
		"permanent_address",
		"permanent_accommodation_type",
		"person_to_be_contacted",
		"emergency_phone_number",
		"relation",
		"bio",
		"unsubscribed",
	]);
	const BREAK_TYPES = new Set([
		"Section Break",
		"Column Break",
		"Tab Break",
		"HTML",
		"Button",
		"Heading",
		"Fold",
	]);
	const EMPLOYEE_FILTER_DOCTYPES = {
		Employee: "name",
		"Leave Application": "employee",
		Attendance: "employee",
		"Attendance Request": "employee",
		"Shift Request": "employee",
		"Expense Claim": "employee",
		"Employee Advance": "employee",
		"Travel Request": "employee",
		"Leave Encashment": "employee",
		"Overtime Request": "employee",
		"Salary Slip": "employee",
		"Salary Structure Assignment": "employee",
		Appraisal: "employee",
		Goal: "employee",
		"Employee Performance Feedback": "employee",
		"QD Performance Improvement Plan": "employee",
		"QD Training Request": "employee",
		"QD Training Nomination": "employee",
		"QD Training Enrollment": "employee",
		"QD Training Attendance": "employee",
		"QD Training Assessment": "employee",
		"QD Training Certification": "employee",
		"QD Grievance": "employee",
		"QD Complaint": "employee",
		"QD Employee Asset Assignment": "employee",
		"QD Asset Loss Damage Case": "employee",
		"QD Asset Recovery": "employee",
		"QD Employee Request": "employee",
		"Employee Separation": "employee",
		"QD Exit Clearance": "employee",
		"Appointment Letter": "employee",
		"Employee Grievance": "employee",
		"Compensatory Leave Request": "employee",
		"Employee Tax Exemption Declaration": "employee",
		"Training Feedback": "employee",
		"Employee Referral": "employee",
		Timesheet: "employee",
		"Employee Checkin": "employee",
		"QD Employee Document": "employee",
		"QD Policy Acknowledgement": "employee",
		"Employee Employment History": "employee",
		"Employee Promotion": "employee",
		Asset: "custodian",
	};
	const HIDE_ON_CREATE = Object.keys(EMPLOYEE_FILTER_DOCTYPES).filter(
		(dt) => dt !== "Employee" && EMPLOYEE_FILTER_DOCTYPES[dt] !== "name"
	);

	frappe.provide("qd_hrms.self_service");

	function isSelfService() {
		if (frappe.boot && frappe.boot.qd_is_self_service) return true;
		const roles = frappe.user_roles || [];
		if (
			roles.includes("System Manager") ||
			roles.includes("HR Manager") ||
			roles.includes("HR User") ||
			roles.includes("Payroll Manager")
		) {
			return false;
		}
		return roles.includes("Employee Self Service") || roles.includes("Employee");
	}

	qd_hrms.self_service.is_self_service_user = isSelfService;

	function ownEmployee() {
		return (frappe.boot && frappe.boot.employee) || null;
	}

	function applyOwnFilter(listview) {
		if (!isSelfService()) return;
		const employee = ownEmployee();
		if (!employee) return;
		const field = EMPLOYEE_FILTER_DOCTYPES[listview.doctype];
		if (!field) return;
		listview.filters = listview.filters || [];
		const exists = listview.filters.some((f) => f[1] === field);
		if (exists) return;
		listview.filters.push([listview.doctype, field, "=", employee]);
	}

	function patchListView() {
		if (!frappe.views || !frappe.views.ListView || frappe.views.ListView.prototype._qdEssPatched) {
			return;
		}
		const orig = frappe.views.ListView.prototype.setup_defaults;
		frappe.views.ListView.prototype.setup_defaults = function () {
			const result = orig.apply(this, arguments);
			applyOwnFilter(this);
			return result;
		};
		frappe.views.ListView.prototype._qdEssPatched = true;
	}

	function lockEmployeeForm(frm) {
		(frm.meta.fields || []).forEach((df) => {
			if (BREAK_TYPES.has(df.fieldtype) || CONTACT_FIELDS.has(df.fieldname)) return;
			frm.set_df_property(df.fieldname, "read_only", 1);
		});
		frm.disable_save = false;
		["History", "Reporting", "Salary"].forEach((group) => {
			if (frm.page && frm.page.clear_custom_actions_of_group) {
				try {
					frm.page.clear_custom_actions_of_group(group);
				} catch (e) {
					/* ignore */
				}
			}
		});
	}

	function bindForms() {
		if (window._qdEssFormsBound) return;
		window._qdEssFormsBound = true;

		HIDE_ON_CREATE.forEach((dt) => {
			const field = EMPLOYEE_FILTER_DOCTYPES[dt];
			frappe.ui.form.on(dt, {
				onload(frm) {
					if (!isSelfService()) return;
					const employee = ownEmployee();
					if (frm.is_new() && employee && frm.fields_dict[field] && !frm.doc[field]) {
						frm.set_value(field, employee);
					}
					if (frm.fields_dict[field]) {
						frm.set_df_property(field, "read_only", 1);
						if (frm.is_new()) {
							frm.set_df_property(field, "hidden", 1);
						}
					}
				},
			});
		});

		frappe.ui.form.on("Employee", {
			refresh(frm) {
				if (!isSelfService()) return;
				lockEmployeeForm(frm);
			},
		});

		frappe.ui.form.on("QD Employee Document", {
			onload(frm) {
				if (frm.is_new() && !isSelfService()) {
					frm.set_value("issued_by_hr", 1);
				}
				if (isSelfService()) {
					frm.set_df_property("issued_by_hr", "read_only", 1);
					frm.set_df_property("employee", "read_only", 1);
				}
			},
		});
	}

	function boot() {
		patchListView();
		if (frappe.ui && frappe.ui.form) {
			bindForms();
		}
	}

	if (document.readyState === "loading") {
		document.addEventListener("DOMContentLoaded", boot);
	} else {
		boot();
	}

	document.addEventListener("page-change", patchListView);
})();

File: qd_hrms_app/qd_hrms/setup/self_service.py
"""My HR workspace, ESS User Type extras, and self-service field locks."""

from __future__ import annotations

import json

import frappe
from frappe.custom.doctype.property_setter.property_setter import make_property_setter
from frappe.permissions import add_permission, update_permission_property

from qd_hrms.setup.bank_tax_pension import SENSITIVE_PERMLEVEL, SENSITIVE_ROLES

WORKSPACE_NAME = "My HR"
ESS_USER_TYPE = "Employee Self Service"

# Extra ESS DocTypes appended to HRMS defaults (never replace the shipped list).
ESS_EXTRA_DOCTYPES = {
	"Attendance": ["read"],
	"Appraisal": ["read", "write"],
	"Goal": ["read", "write", "create"],
	"Employee Performance Feedback": ["read", "write", "create", "submit"],
	"Energy Point Log": ["read"],
	"QD Performance Improvement Plan": ["read"],
	"KRA": ["read"],
	"Appraisal Cycle": ["read"],
	"Appraisal Template": ["read"],
	"Training Event": ["read"],
	"QD Training Course": ["read"],
	"Training Program": ["read"],
	"QD Training Request": ["read", "write", "create"],
	"QD Training Nomination": ["read"],
	"QD Training Enrollment": ["read"],
	"QD Training Attendance": ["read"],
	"QD Training Assessment": ["read"],
	"QD Training Certification": ["read"],
	"QD License Type": ["read"],
	"QD Employee License": ["read"],
	"QD Grievance": ["read", "write", "create", "submit"],
	"QD Complaint": ["read", "write", "create", "submit"],
	"Appointment Letter": ["read"],
	"Asset": ["read"],
	"QD Employee Asset Assignment": ["read"],
	"QD Asset Loss Damage Case": ["read", "write", "create"],
	"QD Asset Recovery": ["read"],
	"QD Employee Request Type": ["read"],
	"QD Employee Request": ["read", "write", "create"],
	"Employee Separation": ["read"],
	"QD Exit Clearance": ["read"],
	"Employee Employment History": ["read"],
	"Employee Promotion": ["read"],
	"Salary Structure Assignment": ["read"],
	"Leave Encashment": ["read", "write", "create"],
	"Overtime Request": ["read", "write", "create"],
	"QD Employee Document": ["read", "write", "create"],
	"QD Policy Acknowledgement": ["read", "write", "submit"],
	"QD Policy": ["read"],
}

SENSITIVE_EMPLOYEE_FIELDS = ("ctc", "salary_mode")

SHORTCUTS = (
	("My Profile", "Employee", "Blue"),
	("My Documents", "QD Employee Document", "Orange"),
	("My Leave", "Leave Application", "Blue"),
	("My Attendance", "Attendance", "Blue"),
	("My Payslips", "Salary Slip", "Orange"),
	("My Assets", "QD Employee Asset Assignment", "Grey"),
	("My Requests", "QD Employee Request", "Blue"),
	("My Performance", "Appraisal", "Blue"),
	("My Training", "Training Event", "Grey"),
	("My Licenses", "QD Employee License", "Orange"),
	("My Letters", "Appointment Letter", "Grey"),
	("My Cases", "Employee Grievance", "Orange"),
)

REQUEST_LINKS = (
	"Attendance Request",
	"Shift Request",
	"Expense Claim",
	"Employee Advance",
	"Travel Request",
	"Leave Encashment",
	"Overtime Request",
	"QD Training Request",
	"QD Employee Request",
)

MORE_LINKS = (
	"Compensatory Leave Request",
	"Employee Tax Exemption Declaration",
	"Training Feedback",
	"Employee Referral",
	"Timesheet",
	"Employee Checkin",
	"Goal",
	"Employee Performance Feedback",
	"Energy Point Log",
	"QD Training Enrollment",
	"QD Training Certification",
	"QD Employee License",
	"QD Grievance",
	"QD Complaint",
	"QD Asset Loss Damage Case",
	"QD Asset Recovery",
	"Employee Separation",
	"QD Exit Clearance",
)

DOCUMENT_LINKS = (
	"QD Employee Document",
	"QD Policy Acknowledgement",
)


def run():
	extend_ess_user_type()
	lock_sensitive_employee_fields()
	ensure_workspace()
	frappe.clear_cache()
	return {
		"workspace": WORKSPACE_NAME,
		"ess_user_type": ESS_USER_TYPE if frappe.db.exists("User Type", ESS_USER_TYPE) else None,
		"ess_extra": [dt for dt in ESS_EXTRA_DOCTYPES if frappe.db.exists("DocType", dt)],
		"sensitive_fields": list(SENSITIVE_EMPLOYEE_FIELDS),
	}


def extend_ess_user_type():
	if not frappe.db.exists("User Type", ESS_USER_TYPE):
		return
	from hrms.setup import append_docperms_to_user_type

	doc = frappe.get_doc("User Type", ESS_USER_TYPE)
	doc.set(
		"user_doctypes",
		[row for row in doc.user_doctypes if row.document_type != "QD Recognition Award"],
	)
	docperms = {dt: perms for dt, perms in ESS_EXTRA_DOCTYPES.items() if frappe.db.exists("DocType", dt)}
	append_docperms_to_user_type(docperms, doc)
	doc.flags.ignore_links = True
	doc.save(ignore_permissions=True)


def lock_sensitive_employee_fields():
	for fieldname in SENSITIVE_EMPLOYEE_FIELDS:
		if not frappe.get_meta("Employee").has_field(fieldname):
			continue
		frappe.db.delete(
			"Property Setter",
			{
				"doc_type": "Employee",
				"field_name": fieldname,
				"property": "permlevel",
			},
		)
		make_property_setter(
			"Employee",
			fieldname,
			"permlevel",
			str(SENSITIVE_PERMLEVEL),
			"Int",
			validate_fields_for_doctype=False,
			is_system_generated=False,
		)
	for role in SENSITIVE_ROLES:
		if not frappe.db.exists("Role", role):
			continue
		if not frappe.db.exists(
			"Custom DocPerm",
			{
				"parent": "Employee",
				"role": role,
				"permlevel": SENSITIVE_PERMLEVEL,
				"if_owner": 0,
			},
		):
			add_permission("Employee", role, SENSITIVE_PERMLEVEL)
		update_permission_property("Employee", role, SENSITIVE_PERMLEVEL, "read", 1)
		update_permission_property("Employee", role, SENSITIVE_PERMLEVEL, "write", 1)


def ensure_workspace():
	shortcuts = [(label, dt, color) for label, dt, color in SHORTCUTS if frappe.db.exists("DocType", dt)]
	request_links = [dt for dt in REQUEST_LINKS if frappe.db.exists("DocType", dt)]
	more_links = [dt for dt in MORE_LINKS if frappe.db.exists("DocType", dt)]
	document_links = [dt for dt in DOCUMENT_LINKS if frappe.db.exists("DocType", dt)]

	if frappe.db.exists("Workspace", WORKSPACE_NAME):
		doc = frappe.get_doc("Workspace", WORKSPACE_NAME)
		doc.shortcuts = []
		doc.links = []
		doc.roles = []
	else:
		doc = frappe.new_doc("Workspace")
		doc.label = WORKSPACE_NAME

	doc.title = WORKSPACE_NAME
	doc.module = "QD HRMS"
	doc.icon = "assign"
	doc.public = 1
	doc.is_hidden = 0
	doc.hide_custom = 0
	doc.parent_page = ""
	if doc.meta.has_field("sequence_id") and not doc.sequence_id:
		doc.sequence_id = 2

	for role in ("Employee", "Employee Self Service"):
		if frappe.db.exists("Role", role):
			doc.append("roles", {"role": role})

	for label, dt, color in shortcuts:
		doc.append(
			"shortcuts",
			{
				"type": "DocType",
				"link_to": dt,
				"doc_view": "List",
				"label": label,
				"color": color,
			},
		)

	_append_card(doc, "My Requests", request_links)
	_append_card(doc, "More Self Service", more_links)
	_append_card(doc, "My Documents", document_links)

	content = [
		_block("header", {"text": '<span class="h4"><b>My HR</b></span>', "col": 12}),
	]
	for label, _dt, _color in shortcuts:
		content.append(_block("shortcut", {"shortcut_name": label, "col": 3}))
	content.append(_block("spacer", {"col": 12}))
	content.append(
		_block("header", {"text": '<span class="h4"><b>Requests &amp; documents</b></span>', "col": 12})
	)
	if request_links:
		content.append(_block("card", {"card_name": "My Requests", "col": 4}))
	if more_links:
		content.append(_block("card", {"card_name": "More Self Service", "col": 4}))
	if document_links:
		content.append(_block("card", {"card_name": "My Documents", "col": 4}))
	doc.content = json.dumps(content)

	doc.flags.ignore_links = True
	doc.flags.ignore_permissions = True
	was_install = frappe.flags.in_install
	frappe.flags.in_install = True
	try:
		doc.save(ignore_permissions=True)
	finally:
		frappe.flags.in_install = was_install


def _append_card(doc, card_name, doctypes):
	if not doctypes:
		return
	doc.append(
		"links",
		{
			"type": "Card Break",
			"label": card_name,
			"hidden": 0,
			"link_count": len(doctypes),
		},
	)
	for dt in doctypes:
		doc.append(
			"links",
			{
				"type": "Link",
				"link_type": "DocType",
				"link_to": dt,
				"label": dt,
				"hidden": 0,
				"is_query_report": 0,
				"onboard": 0,
			},
		)


def _block(block_type, data):
	return {"id": frappe.generate_hash()[:10], "type": block_type, "data": data}
